src/object/cli_obj.c

```diff
@@ -3788,9 +3788,14 @@ obj_comp_cb(tse_task_t *task, void *data)
 	    DAOS_FAIL_CHECK(DAOS_DTX_NO_RETRY))
 		obj_auxi->io_retry = 0;
 
+	/* If map_ver_reply == 0, it means the req never get to the server, so
```

Comment:
In current map_refresh design and API, if the request never got to the server, 0 should be passed in. So this is not the bug.

Reply:
See the API doc of dc_pool_create_map_refresh_task, please:

     *   1 Check if there is a pool map version > the cached version, and if there
     *     is, get it. In this case, pass 0 in \a map_version.

Would you like me to change the description for \a map_version to something like "new known map version"? Or, any better word you can think of?

Reply:
"server pool map version"?

Reply:
So mra_map_version task is the newest map version client knowing?  if that is the case, then if the RPC never get to the server,  it should use the request map version, which is  known by client. 

But could you please explain what is the dp_map_version_known, client known pool map version? but it is already cached in dc_pool.  Or only for refresh task, then why put it in dc_pool, instead into refresh task?  

Hmm, even the refresh task does not get to the right version which the client require, it can still be found by the following requests, besides this is really rare, or I miss sth here?

Reply:
> So mra_map_version task is the newest map version client knowing? if that is the case, then if the RPC never get to the server, it should use the request map version, which is known by client.

mra_map_version is

- server pool map version, which is > 0, if the client has got a server reply, or
- 0, if the client has not heard from a server.

This is defined by the API doc I mentioned above.

> But could you please explain what is the dp_map_version_known, client known pool map version?

dp_map_version_known is the highest pool map version that the client has heard of. E.g., initially we have:

    dp_map is version 10
    dp_map_version_known is 10

then the client receives a server reply, "-DER_STALE, mine is 11":

    dp_map is version 10
    dp_map_version_known is 11

which means the client _has_ version 10, but _knows_ that there is version 11. "My cache is known to be stale."

> but it is already cached in dc_pool. Or only for refresh task, then why put it in dc_pool, instead into refresh task?

This known version is put in dc_pool for the handling of concurrent map_refresh tasks. See [this](https://daosio.atlassian.net/browse/DAOS-8314?focusedCommentId=43335) and later comments on DAOS-8314, please.

> Hmm, even the refresh task does not get to the right version which the client require, it can still be found by the following requests, besides this is really rare, or I miss sth here?

It's hard to say for sure if this is going to be really rare. Probably.

Reply:
>which means the client has version 10, but knows that there is version 11. "My cache is known to be stale.

The only thing client needs to know at the moment is to refresh the pool map. what is the benefit for client to know what is the server pool map version before refresh?   I can understand only one map refresh task is allowed to communicate to the server, but maintaining the map version for refresh task seems a bit over complicate.  

>It's hard to say for sure if this is going to be really rare. Probably.

If this is not true, then there is a bigger problem needs to be addressed IMHO.

Reply:
> The only thing client needs to know at the moment is to refresh the pool map. what is the benefit for client to know what is the server pool map version before refresh?

There are cases where the client is told by the server what pool map version it must acquire. In these cases, if after the client performs a TGT_QUERY_MAP and gets 10, but it knows at least 11 is required, then it can retry TGT_QUERY_MAP on a random server, without retrying the actual I/O.

For the record, I'm against merging this PR for the reasons explained above.

Reply:
I mean I do not see why the server has to tell the client what is the version it will get. I  mean even after the server tell the client, it can still change. 

So what is the point maintaining this on the client side?